rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/WorkerBookKeeper.java

```diff
@@ -54,13 +54,33 @@ public void handleHeartbeat(String workerHostname)
    */
   private void startHeartbeatService(Configuration conf)
   {
-    try {
-      this.heartbeatService = new HeartbeatService(conf);
-      heartbeatService.startAsync();
-    }
-    catch (TTransportException e) {
-      log.fatal("Could not start heartbeat service", e);
+    final int retryInterval = CacheConfig.getServiceRetryInterval(conf);
```

Comment:
The main problem here is we are creating the bookkeeper client in the constructor of HeartbeatService. So in case when there is an exception while creating the bookkeeper client, we are catching the exception in `startHeartbeatService`

What we should do is to move the creation of bookkeeper client in the `runOneIteration` method and create the client only once. In that case `runOneIteration` takes care of the retry part. So even if the client doesn't get created at first attempt, it will retry at regular interval defined by `heartbeatInterval`

Reply:
Makes sense. I've moved the creation of the client to `runOneIteration`, and the change is working on QiB.